documentation/projects/proposals/trust_and_safety/content_report_moderation/20240122-implementation_plan_bulk_moderation_actions.md

```diff
@@ -0,0 +1,526 @@
+
+This action takes a queryset of `DeletedMedia` records, rather than
+`AbstractMedia`. Like `SensitiveMedia`, each `DeletedMedia` record has a related
+`media_obj`, which will be added to the ModerationDecision. The action creates a
+ModerationDecision with type `reversed_deindex` and adds the related records.
+Then it deletes the `DeletedMedia` records.
+
+It is extremely important to note that this action does not immediately result
+in deindexed records being added back to the DB and elasticsearch indices. When
+the records were originally deindexed, they were deleted from both, and there is
+no quick way to restore them without running a data refresh. The confirmation
+page should contain a highly visible `info` block explaining this to the user.
+
+Although the `DeletedMedia` records are deleted, the ModerationDecisions
+preserve the history of these actions.
+
+### Other actions
+
+We do _not_ need to add any new action types for `rejected_reports` and
+`duplicate_reports`, which are only applicable to the Report view flow described
+in the related IP.
+
+## Views
+
+We will add one new read-only view for viewing ModerationDecisions, but most
+changes will be to existing views.
+
+### Media views
+
+We will add the custom bulk actions for `deindex_sensitive`,
+`deindex_copyright`, and `mark_sensitive` to the dropdown in the Media views.
+Bulk moderation decisions of these types will be created by taking these
+actions.
+
+However the current Media views only allow filtering records by uuid. To
+facilitate selecting records for bulk action, we'll update the `search_fields`
+for the main searchbar to search text fields (title, description, tags). We'll
+also add some additional text filters to the sidebar:
+
+- `uuid` - it may still be necessary to search by uuid, so we should preserve
+  this functionality. We include it in a separate filter (rather than keeping it
+  in `search_fields`) for search efficiency.
+- `provider`
+- `creator` - this field should have helptext reminding the user to also filter
+  by provider in order to disambiguate a specific creator.
+
+Additional or more sophisticated filters may be added in the future as
+appropriate.
+
+### DeletedMedia List View
+
+We'll add the `reversed_deindex` action to this list view, allowing maintainers
+to select a list of deindexed records and undo deindexing from this view.
+
+To facilitate easily undoing a particular bulk ModerationDecision, we will add a
+text filter for searching by ModerationDecision id. That allows you to select
+all records that were deleted as part of a particular decision.
+
+### SensitiveMedia List View
+
+We will add the `reversed_mark_sensitive` action to this list view, allowing
+maintainers to select a list of sensitive media and unmark them as sensitive
+from this view.
+
+Similar to the DeletedMedia list view, we will add a filter for
+ModerationDecision id to facilitate easily undoing an entire bulk
+ModerationDecision.
```

Comment:
I guess the importance of these two views being distinct from viewing individual records related to moderation decisions, is that this makes it easier for moderators to reverse a subset or even cross-section of decisions from one or many moderation decisions? Like from these views, a single reversal decision can be used to reverse deindexing caused by multiple different moderation decisions (e.g., a one-off and a bulk decision)?

Just trying to make sure I understand the purpose of these views as opposed to only being able to view the moderation decisions and then within the moderation decision selecting a subset of decisions to reverse.

I suppose the other advantage is that these two views only ever have _current_ decisions, rather than the moderation decisions list which will have current and historical decisions.

Reply:
Basically what you say here, yep. TL;DR but my initial plan _was_ to support reversing decisions from the ModerationDecision list view, but I ended up spending a ton of time trying to think through complexity and edge cases caused by that approach -- which are entirely eliminated by just implementing the reverse from these views. 

I talked about this a little in the Alternatives section, but some more context: it is important to track when a decision has been reversed, so that it cannot be reversed twice -- which is not so easy without complicating the model. I came up with a lot of edge cases along the lines of:
  * Decision A deindexes a set of records. Decision B undoes the deindex. Decision C deindexes a set of records, some of which were affected by Decision A. Then you try to undo Decision A again, which results in **some** of Decision _C_ being 'undone', and the history doesn't really make sense.
  * Decision A deindexes all of Creator X's records, and Decision B undoes this. Decision C deindexes this creator again at a later date, when a larger number of records by Creator X have been ingested since the last time they were deindexed. Decision D is mistakenly made as a reverse of Decision A instead of Decision C, and not all of Creator X's records get reindexed.
  * and so on 😵 

It's also more difficult to reverse subsets -- for example, if you've reversed a subset of a decision, should you be able to later reverse the remaining records? What does that look like? And it's actually impossible to reverse cross-sections, which I could easily imagine wanting when a bulk decision is prompted by a number of individual decisions by moderators for the same creator -- it would be much cleaner to undo all of those in one decision.